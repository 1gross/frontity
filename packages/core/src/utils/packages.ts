import { NormalizedSettings } from "@frontity/file-settings";
import { Package } from "@frontity/types";
import deepmerge from "deepmerge";

// Remove some characters present in the npm package name to
// turn it into a variable name.
export const getVariable = (pkg: string, mode: string) => {
  return (
    pkg
      .replace(/^@/, "")
      .replace(/-/g, "_")
      .replace(/\//g, "__")
      .replace(/\./g, "___") + `_${mode}`
  );
};

// Merge all packages together in a single config that can be passed
// to createStore.
export const mergePackages = ({
  packages,
  state
}: {
  packages: {
    [name: string]: Package;
  };
  state: {
    frontity: {
      mode: string;
      packages: string[];
    };
  };
}): Package => {
  let config: Package = {
    roots: {},
    fills: {},
    state: {},
    actions: {}
  };
  const packageExcludes = state.frontity.packages;
  packageExcludes.forEach(name => {
    const variable = getVariable(name, state.frontity.mode);
    config = deepmerge(config, packages[variable], { clone: false });
  });
  config.state = deepmerge(config.state, state);
  return config;
};
