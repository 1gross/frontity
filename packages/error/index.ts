const isProduction: boolean = process.env.NODE_ENV === "production";
const suffix = "Visit https://community.frontity.org for help! ðŸ™‚";

export const error = (message: string, doThrow = true) => {
  if (!isProduction) {
    if (doThrow) throw new Error(message + suffix);
    console.error(message + suffix);
  } else {
    if (doThrow) throw new Error("Minified error...");
    console.error("Minified error...");
  }
};

export const warn = (message: string) => {
  if (!isProduction) {
    console.warn(message + "\n" + suffix);
  }
};
