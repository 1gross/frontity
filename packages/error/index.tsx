export class FrontityError extends Error {
  constructor(message: string) {
    super(
      `${message}\n\nIf you have any doubts, join our community at https://community.frontity.org/.\n`
    );
  }
}

// Add the `name` property to the class prototype.
FrontityError.prototype.name = "FrontityError";
