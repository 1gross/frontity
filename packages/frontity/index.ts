import create, { events } from "./src/commands/create";

events.on("create", console.log);
create({ name: "random-name" });
