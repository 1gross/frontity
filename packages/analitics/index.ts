import { observe } from "frontity";
import Analytics from "./types";

type AfterCSR = Analytics["actions"]["analytics"]["afterCSR"];
type SendPageview = Analytics["actions"]["analytics"]["sendPageview"];
type SendEvent = Analytics["actions"]["analytics"]["sendEvent"];

// Send a pageview anytime the link changes.
export const afterCSR: AfterCSR = ({ state, actions }) => {
  console.log("afterCSR");

  observe(async () => {
    const { link } = state.router;

    // Wait data to be ready.
    if (state.source.get(link).isReady) {
      /**
       * Execute action in the next tick so we can
       * ensure the title has changed by React.
       */
      setTimeout(() => {
        actions.analytics.sendPageview({
          page: link,
          title: document.title
        });
      });
    }
  });
};

/**
 * Get the functions for every analytics package
 * and run `sendPageview` for each one.
 */
export const sendPageview: SendPageview = ({ state, actions }) => pageview => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendPageview }) => sendPageview(pageview));
};

/**
 * Get the functions for every analytics package
 * and run `sendEvent` for each one.
 */
export const sendEvent: SendEvent = ({ state, actions }) => event => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendEvent }) => sendEvent(event));
};
