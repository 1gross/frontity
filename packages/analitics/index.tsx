import React from "react";
import { connect, Head } from "frontity";
import { Connect } from "frontity/types";
import Analytics from "./types";

type SendPageview = Analytics["actions"]["analytics"]["sendPageview"];
type SendEvent = Analytics["actions"]["analytics"]["sendEvent"];

/**
 * Send a pageview anytime the title changes
 * and data.isReady === true.
 */
export const Root = connect<React.FC<Connect<Analytics>>>(
  ({ state, actions }) => {
    return (
      <Head
        onChangeClientState={newState => {
          if (state.source.get(state.router.link).isReady) {
            actions.analytics.sendPageview({
              page: state.router.link,
              title: newState.title
            });
          }
        }}
      />
    );
  }
);

/**
 * Get the functions for every analytics package
 * and run `sendPageview` for each one.
 */
export const sendPageview: SendPageview = ({ state, actions }) => pageview => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendPageview }) => sendPageview(pageview));
};

/**
 * Get the functions for every analytics package
 * and run `sendEvent` for each one.
 */
export const sendEvent: SendEvent = ({ state, actions }) => event => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendEvent }) => sendEvent(event));
};
