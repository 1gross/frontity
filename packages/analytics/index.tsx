import React from "react";
import { connect, Head } from "frontity";
import { Connect } from "frontity/types";
import Analytics from "./types";

type SendPageview = Analytics["actions"]["analytics"]["sendPageview"];
type SendEvent = Analytics["actions"]["analytics"]["sendEvent"];

/**
 * Send a pageview anytime the title changes
 * and data.isReady === true.
 */
export const Root = connect<React.FC<Connect<Analytics>>>(
  ({ state, actions }) => {
    const { link } = state.router;
    const { isReady } = state.source.get(link);

    // Store the previous title.
    const [prevTitle, setPrevTitle] = React.useState("");

    // Store if a pageview has been sent for this link.
    const [isPageviewSent, setIsPageviewSent] = React.useState(false);

    // Every time link changes, we reset the isPageviewSent value.
    React.useEffect(() => {
      setIsPageviewSent(false);
    }, [link]);

    return (
      <Head
        /**
         * This `titleTemplate` changes the title while
         * `data.isReady` is `false`. This ensures a title change
         * when `data.isReady` becomes `true` but the title hasn't changed.
         */
        titleTemplate={!isReady ? "%s " : ""}
        onChangeClientState={({ title }) => {
          if (isReady && !isPageviewSent && prevTitle !== title) {
            // Store current title.
            setPrevTitle(title);
            // Send pageview.
            actions.analytics.sendPageview({
              page: state.router.link,
              title
            });
            // Mark pageview as sent.
            setIsPageviewSent(true);
          }
        }}
      />
    );
  }
);

/**
 * Get the functions for every analytics package
 * and run `sendPageview` for each one.
 */
export const sendPageview: SendPageview = ({ state, actions }) => pageview => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendPageview }) => sendPageview(pageview));
};

/**
 * Get the functions for every analytics package
 * and run `sendEvent` for each one.
 */
export const sendEvent: SendEvent = ({ state, actions }) => event => {
  Object.keys(state.analytics.namespaces)
    .map(ns => actions[ns])
    .forEach(({ sendEvent }) => sendEvent(event));
};
