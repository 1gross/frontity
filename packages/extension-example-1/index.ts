import { Package, Derived, Action } from "frontity/types";
import Router from "@frontity/router";

interface ExtensionExample1 extends Package {
  name: "@frontity/extension-example-1";
  state: {
    extension1: {
      prop1: number;
    };
    comments: {
      prop2: number;
      prop3: Derived<ExtensionExample1, number>;
    };
    router?: Router["state"]["router"];
  };
  actions: {
    extension1: {
      init: Action<ExtensionExample1>;
      beforeSSR: Action<ExtensionExample1>;
      beforeCSR: Action<ExtensionExample1>;
      afterSSR: Action<ExtensionExample1>;
      afterCSR: Action<ExtensionExample1>;
      action1: Action<ExtensionExample1>;
      action2: Action<ExtensionExample1, number>;
    };
    router?: Router["actions"]["router"];
  };
  roots: {
    extension1: React.ReactType;
  };
  fills: {
    extension1: React.ReactType;
  };
  libraries: {
    comments: {
      Comment: React.ReactType;
    };
  };
}

export default ExtensionExample1;
