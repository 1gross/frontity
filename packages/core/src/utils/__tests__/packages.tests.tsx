import React from "react";
import { getVariable, mergePackages } from "../packages";

describe("getVariable", () => {
  it("should generate different variable names for different packages", () => {
    expect(getVariable("@org/package", "mode")).not.toBe(
      getVariable("org-package", "mode")
    );
    expect(getVariable("@org/package", "mode")).not.toBe(
      getVariable("org.package", "mode")
    );
    expect(getVariable("org-package", "mode")).not.toBe(
      getVariable("org.package", "mode")
    );
    expect(getVariable("@org/package", "html")).not.toBe(
      getVariable("@org/package", "amp")
    );
  });
});

describe("mergePackages", () => {
  const state = {
    frontity: {
      mode: "html",
      packages: ["package-1", "package-2"]
    }
  };
  const packages = {
    package_1_html: {
      roots: {
        namespace1: () => <div>"namespace1"</div>,
        namespace2: () => <div>"namespace2"</div>
      },
      state: {
        namespace1: {
          prop1: "prop1"
        },
        namespace2: {
          prop2: "prop2"
        }
      }
    },
    package_2_html: {
      roots: {
        namespace3: () => <div>"namespace3"</div>
      },
      state: {
        namespace1: {
          prop1: "prop1 overwritten by package_2"
        },
        namespace3: {
          prop3: "prop3"
        }
      }
    }
  };

  it("should output a merged packages", () => {
    expect(mergePackages({ packages, state })).toMatchSnapshot();
  });
});
