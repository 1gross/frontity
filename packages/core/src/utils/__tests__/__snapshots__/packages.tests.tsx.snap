// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergePackages should output a merged packages 1`] = `
Object {
  "actions": Object {},
  "fills": Object {},
  "roots": Object {
    "namespace1": [Function],
    "namespace2": [Function],
    "namespace3": [Function],
  },
  "state": Object {
    "frontity": Object {
      "mode": "html",
      "packages": Array [
        "package-1",
        "package-2",
      ],
    },
    "namespace1": Object {
      "prop1": "prop1 overwritten by package_2",
    },
    "namespace2": Object {
      "prop2": "prop2",
    },
    "namespace3": Object {
      "prop3": "prop3",
    },
  },
}
`;
